packages/core/src/Primitive.tsx

```diff
@@ -0,0 +1,40 @@
+import React from 'react';
+
+export type PrimitiveOwnProps<
+  E extends React.ElementType = React.ElementType
+> = {
+  as?: E;
+  /**
+   * Class names to be appended to the className prop of the component
+   */
+  className?: string;
+  /**
+   * A selection ID used for testing purposes applied as a data attribute
+   * (data-test-id)
+   */
+  testId?: string;
+};
+
+export type PrimitiveProps<E extends React.ElementType> = PrimitiveOwnProps<E> &
+  Omit<React.ComponentProps<E>, keyof PrimitiveOwnProps>;
+
+export type PolymorphicComponentProps<E extends React.ElementType, P> = P &
```

Comment:
Can we use `React.PropsWithoutRef` here? It'd be awesome if we could avoid having to do `ref: typeof otherProps.ref` too.

Reply:
`ref: typeof otherProps.ref` actually helps to enforce a proper reference type based on `as` tag